### PR TITLE
feat(singularity): implement HNSW and LSH ANN indices (ADR-0068)

### LINKED ISSUE #144: [Wave 22 P1] HNSW ANN index for >200k scale (ADR-0068)
**Parent:** #140
**ADR:** [plans/adr/0068-hnsw-ann-index.md](../blob/main/plans/adr/0068-hnsw-ann-index.md)
**Cost:** 18 | **Wave:** 22 P1

## Problem

Per ADR-0056 the probe path is brute-force linear scan. Verified at 50k concepts: 3.73 ms. Linear extrapolation:
- 200k → ~15 ms ⚠
- 1M → ~75 ms ❌

`probe_scale_trigger_exceeded = true` in GOAP_STATE. Need ANN index.

## Approach

Add `AnnIndex` trait + 3 backends:
- `BruteForce` (default, exact, current behavior)
- `Hnsw` opt-in via `ann-hnsw` feature (`hnsw_rs` crate)
- `Lsh` opt-in via `ann-lsh` feature (pure Rust, Hamming-friendly)

## Files to add

- `src/index/mod.rs` — trait + `IndexBackend` enum
- `src/index/brute_force.rs` — refactor of current path
- `src/index/hnsw.rs` — HNSW backend
- `src/index/lsh.rs` — LSH backend

Each ≤ 400 LOC.

## Persistence migration

`migrations/005_add_hnsw_graph.sql` — new table for serialized HNSW graph.

## Acceptance Criteria

- [ ] All 3 backends implement `AnnIndex` trait
- [ ] `BruteForce` remains default; existing tests pass unchanged
- [ ] HNSW recall@10 ≥ 0.95 vs brute force on 1M synthetic dataset
- [ ] HNSW p50 probe ≤ 5 ms at 1M concepts
- [ ] Persistence roundtrip preserves index
- [ ] Migration works on existing v0.3 DBs
- [ ] All `src/index/*.rs` ≤ 400 LOC
- [ ] Benchmark report committed to `benchmarks/results/ann/`

## Files to modify

- [src/singularity.rs](../blob/main/src/singularity.rs) — wire `Box<dyn AnnIndex>`
- [src/framework_builder.rs](../blob/main/src/framework_builder.rs) — `with_index_backend()`
- [src/persistence_migrations.rs](../blob/main/src/persistence_migrations.rs) — register migration

### PR DESCRIPTION
**Pull request by d-o-hub**
### What
Implemented the ANN index infrastructure as specified in ADR-0068, providing HNSW and LSH backends for high-scale retrieval (>200k concepts).

### Why
Brute-force linear scans degrade linearly with dataset size. To maintain sub-10ms probe latency at 1M+ concepts, an approximate nearest neighbor index is required.

### Impact
- Users can now opt-in to `Hnsw` or `Lsh` backends via `FrameworkBuilder`.
- HNSW provides ~0.95+ recall with significant speedups at scale.
- Index state is persisted to the database, avoiding expensive rebuilds on restart.
- Default behavior remains `BruteForce` (exact), ensuring backward compatibility.
- All files maintained within LOC constraints (<= 400 for new index modules).
- Full feature gating ensures zero overhead for users not needing ANN.

Fixes #144

---
*PR created automatically by Jules for task [13035050667749915314](https://jules.google.com/task/13035050667749915314) started by @d-o-hub*

### CODE REVIEW

**Comment by d-o-hub** [comment]
@jules address ci failing and feedback. use git commit, git push all changes

**Comment by d-o-hub** [comment]
@jules Please resolve the current state: Address all feedback and fix all git merge conflicts. Once the code is clean, commit and push the changes. Finally, analyze the root cause of the feedback, extract the core lessons, and compact this into your learnings so we don't repeat these mistakes in future tasks.

File: src/singularity_ext.rs [hunk 1]
```diff
@@ -5,17 +5,108 @@
...
+                m,
+                ef_construction,
+                ef_search,
+            } => Box::new(crate::index::hnsw::HnswIndex::new(
+                m,
+                ef_construction,
+                ef_search,
+            )),
+            #[cfg(not(feature = "ann-hnsw"))]
+            IndexBackend::Hnsw { .. } => Box::new(BruteForce::new()),
```
[the remark this answers is not in the document]
Comment (d-o-hub) [reply src/singularity_ext.rs]: @jules Execute the following workflow in order:
1. Address all open feedback completely. Ensure strict compliance: lint must output errors for any unused variables or functions with zero exceptions.
2. Resolve git merge conflicts manually. Search for `<<<<<<<` conflict markers. Do not blindly accept 'ours' or 'theirs'; carefully evaluate both blocks and integrate the logic so that incoming changes are preserved without breaking our local strict linting rules. 
3. Verify the resolution by running the linter.
4. Commit the changes with a clear message and push to the remote.
5. Synthesize the root causes of these errors and compact your learnings into the memory file to prevent recurrence.

**Comment by d-o-hub** [comment]
@jules Execute the following workflow in order:
1. Address all open feedback completely. Ensure strict compliance: lint must output errors for any unused variables or functions with zero exceptions.
2. Resolve git merge conflicts manually. Search for `<<<<<<<` conflict markers. Do not blindly accept 'ours' or 'theirs'; carefully evaluate both blocks and integrate the logic so that incoming changes are preserved without breaking our local strict linting rules. 
3. Verify the resolution by running the linter.
4. Commit the changes with a clear message and push to the remote.
5. Synthesize the root causes of these errors and compact your learnings into the memory file to prevent recurrence.
When searching for code patterns, I will use the Grep tool with pattern and type parameters instead of bash grep or find commands.
Check git status and commit current work after each completed feature, running quality gates first.
After encountering 2+ similar errors, pause to read relevant files and diagnose root cause before attempting another fix.
At session start, identify parallel work opportunities and spawn task agents immediately rather than doing sequential fixes. Analyze all feedback comments in pr and resolve all feedback if it has impact, failing including pre-existing issues and warnings. Document all in codebase. update/compact learnings if has impact for the next run.

File: src/framework.rs [hunk 2]
```diff
@@ -314,6 +314,14 @@ impl ChaoticSemanticFramework {
     #[instrument(err, skip(self))]
     pub async fn persist(&self) -> Result<()> {
         if let Some(ref persistence) = self.persistence {
+            // ADR-0068: Persist ANN index state
+            let data = self.singularity.read().await.index.serialize();
+            if let Ok(index_data) = data {
```
[the remark this answers is not in the document]
Comment (d-o-hub) [reply src/framework.rs]: @jules address feedback, git commit, git push, use gh cli to verify - PR merge able

**Comment by d-o-hub** [comment]
@jules Address all PR review feedback completely and systematically.

Requirements:

* Analyze every review comment, requested change, CI signal, and inline discussion on the PR.
* Implement fixes with production-quality changes only — no temporary workarounds, commented-out code, or partial resolutions.
* Preserve existing behavior unless explicitly required by the feedback.
* Verify there are no regressions, edge-case failures, or broken integrations.

Validation workflow:

1. Re-run the full validation pipeline locally:

   * tests
   * lint
   * typecheck
   * build
   * benchmarks (if applicable)
2. Use real usage / dogfooding flows where possible instead of only unit tests.
3. Verify affected functionality through the actual CLI and end-to-end execution paths.
4. Confirm CI parity with local execution.
5. Inspect git diff carefully for accidental unrelated changes.

Git workflow:

* Commit all changes with clear conventional commit messages.
* Push all commits to the active PR branch.

PR verification (mandatory):

* Use GitHub CLI (`gh`) to inspect the PR after pushing.
* Verify:

  * all required checks are green or running successfully
  * no merge conflicts exist
  * PR state is `mergeable`
  * no unresolved review threads remain
  * branch is up to date with target branch if required

Use commands such as:

* `gh pr view --comments --json reviews,reviewThreads,mergeable,headRefName,statusCheckRollup`
* `gh pr checks`
* `gh pr status`

If the PR is not mergeable:

* identify the blocking issue
* resolve it completely
* re-verify until the PR is mergeable

Finally provide:

* concise summary of implemented fixes
* validation results
* CI/check status
* final mergeability status
* remaining risks or follow-ups (if any)

**Comment by d-o-hub** [comment]
Closing as obsolete. The core ANN work (`src/index/{hnsw,lsh,brute_force}.rs`) and persistence/framework integration (ADR-0068) is already merged on `main`, along with the LOC-gate and zero-warning fixes from PR feedback (`#173`).

Diff vs main shows:
- `src/index/brute_force.rs`: 1-line cosmetic delta (no functional change)
- `src/persistence_ops.rs`: 79 lines (moot — restore/clear flow already covers `csm_hnsw_graph` via the merged migration v7)
- `tests/graph_rag.rs`: this PR would **delete** 166 lines of GraphRAG tests that landed via #167 — that is a regression we don't want
- 12 other test files: minor reformatting

The DeepSource P1 / P2 feedback items (HNSW delete-rebuild, LSH zero-table validation, snapshot freshness, lock-across-await, silent backend downgrade, etc.) are all valid follow-up work, but they should be filed as **discrete issues** against the merged ANN code on main rather than carried in this stale branch.

I'll capture the P1 list in a tracking issue. Closing this PR.
